Rewrite group_batch_fusion.find_independent_subset_greedy() to be iterative.

aorenste opened the pull request:
Improve performance of inductor searching large graphs for potential fusions.
Also adds some direct unit tests of find_independent_subset_greedy() to ensure that the rewrite didn't break behavior.

Fixes #98467

Previously find_independent_subset_greedy() was recursive and the example from the issue would cause it to blow out the stack. This changes it to be iterative and also caches some of the computed dependencies (it can't cache all of them because the caller is allowed to change the graph during the iteration).

Fusion is still slow - but at least finishes.

After this change the example given in #98467 has the following backend timings (on one particular CPU):
eager timing: 3m:23s
aot_eager timing: 4m:12s
inductor timing: 22m:24s

Possible future work to improve this further:
1. In dynamo limit the amount of inlining allowed before falling back to a graph break. This test ends up tracing through 483k bytecodes generating the graph.
2. In inductor have a limit so we don't exhaustively search the graph for fusion possibilities.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118324



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

aorenste committed:
WIP: fixing recursion depth in group_batch_fusion

[ghstack-poisoned]

aorenste committed:
Update on "WIP: fixing recursion depth in group_batch_fusion"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -740,46 +741,100 @@ def __init__(self, **kwargs):
+    max_fuse_set_size = graph_search_options["max_fuse_set_size"]
+
+    # Cache for the computed dep_sets when we know we're going to need it again.
+    cache: Dict[torch.fx.Node, Set[torch.fx.Node]] = {}
```
Isn't this cache invalid if any fusions happen?

aorenste replied:
I'm not done with this PR yet (it's still in draft).  But yes - do we do that? I would imagine that changing the graph during iteration would cause problems for the original algorithm as well.

peterbell10 replied:
This coroutine yields `subset` back into `apply_group_batch_fusion` which may alter the graph. The current algorithm doesn't carry any state over to the next iteration of the while loop, so I don't see why it would be an issue there.

aorenste replied:
Wouldn't it be a problem if a node in `node_list` was removed from the graph? The old algorithm would still consider that node in its check even though the node is no longer in the graph. If we're just fusing nodes from `subset` that are being returned then I guess that wouldn't be an issue because they've already been removed from `node_list`.

Anyway - you're right that the way the call is being used would cause a problem with the cache - I'll remove it before I publish. Good catch.

peterbell10 replied:
>  If we're just fusing nodes from subset that are being returned then I guess that wouldn't be an issue because they've already been removed from node_list.

This is correct, it only considers `subset`

aorenste committed:
Update on "WIP: fixing recursion depth in group_batch_fusion"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: fixing recursion depth in group_batch_fusion"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: fixing recursion depth in group_batch_fusion"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: fixing recursion depth in group_batch_fusion"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: fixing recursion depth in group_batch_fusion"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: fixing recursion depth in group_batch_fusion"

eager timing: 3:23
aot_eager timing: 4:12
inductor timing: 22:24




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -740,46 +752,105 @@ def __init__(self, **kwargs):
         super().__init__(aten.mul.Tensor, **kwargs)
 
 
+class _OrderedSet:
```
In Python, dicts are now ordered, so you should instead use a Dict[Key, None] for this instead of rolling out your own ordered set.

aorenste replied:
This is just an encapsulation of Dict[Key, None] (via OrderedDict). I would think Isn't it more clear to a reader that the code isn't relying on the dict value by eliminating it?

oulgen replied:
We have quite a lot of this pattern in the codebase. Not the best regex but you can browse: https://fburl.com/biggrep/cp0sy1sg

If we want to proceed with the OrderedSet wrapper, it might be good to unify these, although someone more familiar with the codebase should pitch in here. @ezyang ?

eellison replied:
Yea I think OrderedSet would be an improvement throughout the codebase (especially in compiler). let's leave that for another PR though.

oulgen commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -740,46 +752,105 @@ def __init__(self, **kwargs):
+class _OrderedSet:
+    def __init__(self, param=None):
+        if param:
+            self.rep = OrderedDict({k: None for k in param})
```
we should kill OrderedDict from the codebase now that dicts are ordered

aorenste replied:
Or we could keep it to indicate that the ordering is part of the expected behavior and not just a side-effect...

oulgen commented:
Nice job! Can you also look at improving the 22m compilation time as a follow up to this? @eellison wanted us to add Tacotron back to dynamo benchmarks

eellison commented:
Cool ! Yea I think enabling tacotron2 is a good goal because it's a model in torchbench that has pathological compile time and can guide us towards improvements. It might be an accumulation of bad compile time algorithmic complexity (like here), it might be a lack of larger features (Scan op?  model has LSTM). It doesn't necessarily need to be a top priority but I think will be helpful. Maybe we could do an ablation of where the compile time is spent and then go from there ?

aorenste committed:
Update on "Rewrite group_batch_fusion.find_independent_subset_greedy() to be iterative."

Improve performance of inductor searching large graphs for potential fusions.
Also adds some direct unit tests of find_independent_subset_greedy() to ensure that the rewrite didn't break behavior.

Fixes #98467

Previously find_independent_subset_greedy() was recursive and the example from the issue would cause it to blow out the stack. This changes it to be iterative and also caches some of the computed dependencies (it can't cache all of them because the caller is allowed to change the graph during the iteration).

Fusion is still slow - but at least finishes.

After this change the example given in #98467 has the following backend timings (on one particular CPU):
eager timing: 3m:23s
aot_eager timing: 4m:12s
inductor timing: 22m:24s

Possible future work to improve this further:
1. In dynamo limit the amount of inlining allowed before falling back to a graph break. This test ends up tracing through 483k bytecodes generating the graph.
2. In inductor have a limit so we don't exhaustively search the graph for fusion possibilities.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge
